test(hmr): delete file not used anymore

sapphi-red opened the pull request:
Make HMR work properly when a file is deleted.
This PR adds a test for the case where a file not used anymore is deleted.

sapphi-red commented:
* **#5932** <a href="https://app.graphite.com/github/pr/rolldown/rolldown/5932?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 2  dependent PRs ([#5933](https://github.com/rolldown/rolldown/pull/5933) <a href="https://app.graphite.com/github/pr/rolldown/rolldown/5933?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#7349](https://github.com/rolldown/rolldown/pull/7349) <a href="https://app.graphite.com/github/pr/rolldown/rolldown/7349?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>) 👈 <a href="https://app.graphite.com/github/pr/rolldown/rolldown/5932?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=5932).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sapphi-red commented on crates/rolldown/tests/rolldown/topics/hmr/delete_file_not_used_anymore/artifacts.snap:
```diff
@@ -0,0 +1,170 @@
+# HMR Step 1
+## Errors
+
+### UNLOADABLE_DEPENDENCY
```
Since `child.js` is not imported anymore, we should simply ignore the update call. Or the alternative is to tell Vite not to call `generateHmrPatch` (this requires a way to tell Vite that `child.js` is not imported anymore).

That said, maybe the new API would fix this?

sapphi-red commented on crates/rolldown/tests/rolldown/topics/hmr/delete_file_not_used_anymore/artifacts.snap:
````diff
@@ -0,0 +1,170 @@
+### UNLOADABLE_DEPENDENCY
+
+```text
+[UNLOADABLE_DEPENDENCY] Error: Could not load child.js - 指定されたファイルが見つかりません。 (os error 2).
````
The CI will probably fail because this contains OS language specific error message. I'll add a workaround if we want to merge this PR before fixing this error.

sapphi-red commented on crates/rolldown/tests/rolldown/topics/hmr/delete_file_not_used_anymore/artifacts.snap:
````diff
@@ -0,0 +1,170 @@
+
+```js
+//#region child.js
+var init_child_0 = __rolldown_runtime__.createEsmInitializer((function() {
+	try {
+		var __rolldown_exports__ = {};
+		__rolldown_runtime__.__export(__rolldown_exports__, { value: () => value });
+		__rolldown_runtime__.registerModule("child.js", { exports: __rolldown_exports__ });
+		const hot_child = __rolldown_runtime__.createModuleHotContext("child.js");
+		const value = "child";
+	} finally {}
+}));
+
+//#endregion
````
We don't have to include `child.js` in this patch because `child.js` is not imported. The next patch should include `child.js`.

sapphi-red commented on crates/rolldown/tests/rolldown/topics/hmr/delete_file_not_used_anymore/artifacts.snap:
```diff
@@ -0,0 +1,197 @@
+
+## Meta
+
+- update type: patch
```
TODO: we should make this "step 2" no-op because child.js is not used

sapphi-red replied:
Will do in #7348

sapphi-red commented on crates/rolldown_dev/src/bundle_coordinator.rs:
```diff
@@ -129,20 +130,34 @@ impl BundleCoordinator {
+              for path in batched_event.detail.paths {
+                changed_files.insert(path, WatcherChangeKind::Update);
+              }
+            }
```
TODO: We should handle rename events here

sapphi-red commented on crates/rolldown_dev/src/bundling_task.rs:
```diff
@@ -82,14 +83,11 @@ impl BundlingTask {
-            // FIXME: use proper WatcherChangeKind for created/removed files.
-            .watch_change(changed_file.to_str().unwrap(), WatcherChangeKind::Update)
-            .await?;
+          plugin_driver.watch_change(changed_file.to_str().unwrap(), *event).await?;
```
TODO: We should add tests for each event kinds.

sapphi-red replied:
https://github.com/rolldown/rolldown/pull/7349

sapphi-red commented on crates/rolldown_dev/src/bundling_task.rs:
```diff
@@ -132,8 +130,9 @@ impl BundlingTask {
       .iter()
-      .map(|p| p.to_string_lossy().to_string())
-      .collect::<Vec<_>>();
+      // TODO: handle delete events properly
```
TODO: think about what should be done here

sapphi-red commented on crates/rolldown_dev/src/bundling_task.rs:
```diff
@@ -196,7 +198,8 @@ impl BundlingTask {
     } else {
       ScanMode::Partial(
-        self.input.changed_files().iter().map(|p| p.to_string_lossy().into()).collect(),
+        // TODO: handle delete events properly
```
TODO: think about what should be done here

sapphi-red reviewed (commented):
Still WIP but at least it now passes the tests

sapphi-red commented on crates/rolldown_dev/src/dev_engine.rs:
```diff
@@ -304,22 +306,29 @@ impl DevEngine {
+    &self,
+    changed_files: FxIndexMap<PathBuf, WatcherChangeKind>,
+  ) {
+    for (path, event) in changed_files {
+      // Create a synthetic file change event to simulate real file system changes
+      let notify_event = notify::Event {
+        kind: if event == WatcherChangeKind::Delete {
+          notify::EventKind::Remove(notify::event::RemoveKind::Any)
+        } else {
+          notify::EventKind::Modify(notify::event::ModifyKind::Data(notify::event::DataChange::Any))
+        },
+        paths: vec![path],
+        attrs: notify::event::EventAttributes::default(),
+      };
+
+      let event =
+        rolldown_fs_watcher::FsEvent { detail: notify_event, time: std::time::Instant::now() };
+
+      // Send WatchEvent message to coordinator (simulates real file change)
+      // The coordinator will automatically schedule a build via handle_file_changes
+      let _ = self.coordinator_sender.send(CoordinatorMsg::WatchEvent(Ok(vec![event])));
+    }
```
This should align more with how notify emits the events.

hyf0 commented on crates/rolldown/src/bundler/impl_bundler_hmr.rs:
```diff
@@ -1,15 +1,17 @@
   pub async fn compute_hmr_update_for_file_changes(
     &mut self,
-    changed_file_paths: &[String],
+    changed_file_paths: &FxIndexMap<String, WatcherChangeKind>,
```
(Thought, not blocking)I tend to remove `WatcherChangeKind` but have a extra parameter `removed_file_paths`.

hyf0 commented:
Does the AST of removed modules still exist? If so, I think we need an issue to track it how to clean up properly.

sapphi-red commented:
> Does the AST of removed modules still exist? If so, I think we need an issue to track it how to clean up properly.

I guess so. I'll create an issue that tracks these problems.

sapphi-red committed:
test(hmr): delete file not used anymore (#5932)

Make HMR work properly when a file is deleted.
This PR adds a test for the case where a file not used anymore is deleted.
